sycl/include/CL/sycl/detail/helpers.hpp

```diff
@@ -188,6 +188,19 @@ class Builder {
     item<Dims, false> LocalItem = createItem<Dims, false>(LocalSize, LocalId);
     return createNDItem<Dims>(GlobalItem, LocalItem, Group);
   }
+
+  template <int Dims, bool WithOffset>
+  static auto getElement(item<Dims, WithOffset> *)
+      -> decltype(getItem<Dims, WithOffset>()) {
```

Comment:
Useless line except the '{' ?

Reply:
If I correctly understand the comment you suggest to just use `auto`. If so, I am intentionally leave it with `decltype` to allow it to be compiled with C++11. As far as I know it is still the requirement of the implementation and might be potentially useful.

Reply:
I see... I keep forgetting about this C++11 constraint...

Reply:
I am even more troubled when I read in the title of the PR:

> Allow C++14 Generic lambda expression as the kernel for
> parallel_for_work_group

Are you testing the C++ version to use more or less modern features then?

Reply:
First of all let me go back to your first question. I've just recalled one more reason why I use trailing return type with `decltype`. `getElement` now returns `const item` when the item passed to argument. This behavior was in the implementation before my changes. I would like to preserve that semantics. On the other hand, I don't want to care if the implementers would decide to change this semantics to just `item`. `decltype` did this work for me, just returns what the underlying call returns. Just `auto` (without `decltype`) eliminates `const`

Here is the example: https://godbolt.org/z/2XDnB7

Regarding your second question: This feature works with either C++ (starting from C++11). Internally we just try to take the address of the non-static member function. If we succeed we use that, if not use our default. The is nothing specific from C++14/17 to implement that. On the other hand, if user chooses don't use generic lambda from C++14 it still works. So, users themselves need to choose C++14 or higher to use the generic lambda in their code, but our implementation doesn't rely on C++14 or higher features anyhow.
Another interesting note: Of course we mainly talking about generic lambdas since this is the most important usecase. But at the end nothing prevents users to use the any generic Functor. For example:

```cpp
class Functor {
    template <typename T>
    void operator()(T) const {
        // ...
    }
```
The code above still may be considered as the generic Functor. Generic lambda from C++14 is a just a subset. So, in theory users may even use that features with generic Functors even in C++11. As far as I remember, I intentionally didn't use "lambda" word in the extension.